### DISCUSSION TITLE
[knitr] Options: Chunk options and package options | /knitr/options/

### DISCUSSION THREAD

**Discussion by yihui** [rbind/yihui#746] [Comments]
https://yihui.org/knitr/options/

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Can you try the latest version (0.4)? I do not see the line with 0.4.

> Originally posted on 2012-03-24 05:31:26

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
You almost never need to open and close a graphical device in a code chunk, so please remove windows(10,6) and dev.off(); knitr will open its device to record your plots, but if you have opened another device, knitr will be unable to know new plots produced.

> Originally posted on 2012-04-03 21:45:42

**Reply by yihui** [reply]
If you are using the latest version of knitr and running it under non-interactive mode (e.g. call Rscript), this should not be a problem; knitr will use fig.width and fig.height to open options('device').

If you are in an interactive R session (open R and type code), this might be problematic. It is easy to fix either on my side, though. You can also fix this by yourself by putting something like this in your first chunk:

    options(device = function(...) {
      windows(..., width = opts_current$get('fig.width'), 
              height = opts_current$get('fig.height'))
    })

However, I still recommend you to call R non-interactively.

> Originally posted on 2012-04-04 17:29:57

**Reply by yihui** [reply]
Sorry, I had a typo in the code above; [dev.new](http://dev.new)() should be windows(). For interactive vs non-interactive sessions, see ?interactive in R help.

Anyway, I have fixed this problem in the development version, so you do not really need to do anything but install the latest version from Github: <https://github.com/yihui/knitr> (or wait for my next CRAN release). Thanks!

> Originally posted on 2012-04-04 22:40:51

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Good to know. You should see a warning message about this situation when you call knit().

> Originally posted on 2012-04-08 18:53:44

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Normally you can use echo=FALSE, results='hide' to keep the figures only. If you still find the gray box, you probably need to update the package.

For the large preamble, currently the suboptimal solution is opts_knit$set(self.contained = FALSE).

> Originally posted on 2012-05-15 19:32:26

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Most examples are under <https://github.com/yihui/knitr/tree/master/inst/examples> ; in your case, you can look at knitr-minimal.Rmd or knitr-upload.Rmd; actually these chunk options are exactly the same as the ones you write in Sweave. Instead of <<options>>=, you just write ``` {r options}.

> Originally posted on 2012-05-15 19:13:41

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Glad you figured it out by yourself. All the valid option values are listed in this page, and I have emphasized in the beginning of this page that options must be valid R code.

> Originally posted on 2012-05-25 02:00:37

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
take a look at the corresponding TeX log file under figure/ to see what is wrong there; tikz requires LaTeX expertise, so if you are not familiar with the tikzDevice package, you probably should spend a couple of hours learning it.

> Originally posted on 2012-06-01 20:08:25

**Reply by yihui** [reply]
LyX will not really update your document unless there has been a change in the TeX output; if you only update a plot, the tex code is probably not changed although the plot itself has been changed. In that case, you need to make a small change in any place of your LyX document.

> Originally posted on 2012-06-01 20:12:26

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Because fig.lp is 'fig:' by default, you need to ref{fig:fig1} instead of ref{fig1}. If you want the latter, you can set opts_chunk$set(fig.lp = '') so that no prefix will be added by knitr.

> Originally posted on 2012-07-04 20:55:14

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I have heard other users report the same problem under Ubuntu, and I do not know why it does not work for you. I'm using Ubuntu as well.

There was indeed a small bug before v0.6.3, just in case you were using an older version.

> Originally posted on 2012-07-10 01:12:27

**Reply by yihui** [reply]
You probably need par(bg='white') if you are using base R graphics; I just found that FFmpeg did not work with PNG's with transparent background.

> Originally posted on 2012-07-10 16:07:13

**Reply by yihui** [reply]
MP4 should be very common nowadays. I have no idea about what is wrong. Anyway, you will soon be able to write animations into HTML pages like those in <http://taiyun.github.com/>

> Originally posted on 2012-07-12 03:55:56

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
It is almost impossible to convert animations in HTML to PDF, because they were drastically different in this aspect. That being said, you can create either LaTeX/PDF animations or HTML animations with the same code and syntax ([fig.show](http://fig.show)='animate'), so you'd better do that from the source rather than generating markdown and convert the animations to PDF later.

It is certainly possible to do GIF. Just file a feature request to <https://github.com/yihui/knitr/issues>

In the development version of knitr, you can already make animations with JavaScript (hook_scianimator) or Flash (hook_r2swf), but no instructions have been given. We are working on it. Hopefully you will see some interesting announcements next week.

> Originally posted on 2012-07-10 01:19:47

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I do not understand your question. I need a reproducible example along with your sessionInfo() after library(knitr).

BTW, fig=TRUE has been deprecated long time ago; you do not need fig=TRUE in knitr.

> Originally posted on 2012-07-10 16:05:54

**Reply by yihui** [reply]
Oh, this is not an appropriate place to paste the code. I can barely read it. Could you please file an issue to <https://github.com/yihui/knitr/issues> ? Thanks! And I doubt if the above example is reproducible for me, because I do not have pdata.txt

> Originally posted on 2012-07-10 20:20:23

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I did not mention this in the documentation. Strictly speaking, chunk labels must be quoted as well because they are character strings. In most cases they do not need to be quoted because they are valid R expressions if not being quoted, e.g. a-b, or c_def. However, in your case car-function is not a valid R expression, because "function" is a reserved keyword in R. I will add an explanation to this page later. Thanks for the report!

> Originally posted on 2012-07-13 17:26:35

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
use results='asis'; because you are writing *raw* HTML from R

> Originally posted on 2012-07-18 20:51:52

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
can you finish reading all the 3 bullet points in the beginning of this page?

btw, if you update knitr to v0.7 (which is already on CRAN now), you will see a more detailed note about the reason of the failure

> Originally posted on 2012-07-18 20:53:36

**Reply by yihui** [reply]
I do not know why you failed or why your R chose knitr 0.6.3, since the one on CRAN should be v0.7: <http://lib.stat.cmu.edu/R/CRAN/web/packages/knitr/index.html>

> Originally posted on 2012-07-18 23:55:34

**Reply by yihui** [reply]
Sigh... Could you just scroll down a little bit?

> Originally posted on 2012-07-18 23:58:48

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
That is not tricky. There is nothing to remember. The rule is simple: make sure whatever you write is valid to R; results=asis is not valid because there is no object named "asis". Just think about the difference between function(results=asis) and function(results="asis"). I recommend you to stop spending so much time experimenting by yourself. I have documented this clearly in this page -- read everything from the top up to the line "All built-in options in knitr are:".

> Originally posted on 2012-07-19 21:03:35

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Glad to know you figured it out by yourself :)

> Originally posted on 2012-07-26 23:27:44

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sorry I just noticed the comment was sent to the moderation queue. But apparently there are more people to help you on SO :) <http://stackoverflow.com/q/11668904/559676>

> Originally posted on 2012-07-27 19:58:19

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I saw the same question on SO: <http://stackoverflow.com/q/11782037/559676>

This is pretty weird, and I do not know how this can happen without set.seed(). Anyway, please try to clean up the cache directory or start with a minimal example to diagnose the problem.

One possibility is that you have a chunk which was cached and involved with random number generation before this chunk, so that when this chunk is evaluated, the random seed from that chunk is used. If that is the case, it is not a basic question. It is actually very deep.

> Originally posted on 2012-08-04 23:44:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
yes, use fig.scap='Example'

> Originally posted on 2012-08-13 16:31:50

**Reply by yihui** [reply]
Use fig.cap='textit{abc}'; character options should be quoted, as mentioned in the top of this page.

> Originally posted on 2012-08-15 14:56:26

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
See comments below your SO post: <http://stackoverflow.com/q/12133526/559676>

> Originally posted on 2012-08-26 21:48:22

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I do not know how to set the default font size for base graphics, but in ggplot2 you can set the theme to use a base font size. BTW, the font size also depends on whether you change the size of the whole figure in LaTeX.

> Originally posted on 2012-09-14 03:47:26

**Reply by yihui** [reply]
but split=TRUE is not supposed to have any effects in the font size; it only splits the tex code of an R code chunk into a separate file; if split=TRUE works, split=FALSE (default) should be the same.

> Originally posted on 2012-09-15 03:27:43

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
The description is pretty clear to me, but unfortunately there is no obvious solution. One thing that you might try is to draw the four plots in a single layout, e.g.

```{r fig.height=5*3+3}  
layout(matrix(1:4, 4), heights=c(5, 5, 5, 3))  
par(mar = c(4, 4, .1, .1))  
plot(rnorm(10))  
plot(rnorm(10))  
plot(rnorm(10))  
plot(runif(10))  
```

> Originally posted on 2012-10-19 05:12:03

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Let's focus on the SO post: <http://stackoverflow.com/q/12983930/559676>

> Originally posted on 2012-10-20 01:47:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I do not see any problems with the latest version of knitr and TeXLive 2012. See the attached image as an example. The source document is at <https://github.com/yihui/knitr-examples/blob/master/055-multicols.Rnw>

> Originally posted on 2012-10-29 16:42:59

**Reply by yihui** [reply]
The framed package is really tricky to use: <http://yihui.name/knitr/demo/framed/>

In this case, I'm not aware of any solutions either. Sorry...

> Originally posted on 2012-10-29 22:50:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Thanks! Perhaps I did not make it clear enough: cairo_pdf() is in base R (Paul's web page is a little bit outdated; Windows also has cairo_pdf now); CairoPDF is in the Cairo package, and Cairo_pdf is in cairoDevice.

> Originally posted on 2012-11-05 17:50:54

**Reply by yihui** [reply]
I'm not aware of any existing solutions. None of the Cairo devices above supports CMYK. I have a hack in mind, which is to use the tikz device in the tikzDevice package. You can probably post-process the tex file generated by tikzDevice to replace the RGB colors by CMYK colors.

> Originally posted on 2012-11-07 01:25:09

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
try the width option, e.g. options(width = 50)

> Originally posted on 2012-11-07 20:03:57

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Use the chunk option comment=NA or comment='' if you do not want ##.

For portuguese characters, it is an encoding problem. It is complicated without further information -- where do these characters come from? What was their original encoding? Which editor did you use?

> Originally posted on 2012-12-02 03:11:55

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
So what is the encoding of the database? I suppose there must be an option in RPostgresSQL, but I have never used this package before, so I'm not sure.

Ideally if everything is UTF-8, there will be much less pain.

> Originally posted on 2012-12-02 04:26:33

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Are you using the latest version of knitr? Can you provide a minimal example? I do not see how this could happen:

documentclass{article}  
begin{document}  
<<test, fig.cap='asdf'>>=  
x=rnorm(100)  
plot(x)  
x=rnorm(100)  
plot(x)  
x=rnorm(100)  
plot(x)  
@  
end{document}  

> Originally posted on 2012-12-12 02:13:27

**Reply by yihui** [reply]
By setting fig.cap, you mean to use the figure environment, which in turn means you must make sure the content of the figure environment can fit into one page, which is not true in your case. That is a LaTeX problem.

What I can suggest at most is to make your plots smaller, e.g. fig.width=3, fig.height=3

> Originally posted on 2013-01-09 21:40:52

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
A better place to ask is StackOverflow: <http://stackoverflow.com/questions/tagged/r>

> Originally posted on 2013-01-04 18:14:20

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
You can only use different background colors for different chunks (see the `background` option above), but you cannot really use different highlighting themes for tokens in different chunks. You can only have one highlighting theme in one document.

> Originally posted on 2013-01-17 02:22:37

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
FAQ 1: <http://bit.ly/knitr-faq>

> Originally posted on 2013-01-18 19:21:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
no, but you can hack the source, e.g. gsub('^<<', '<<eval=FALSE,', ...)

> Originally posted on 2013-02-13 19:13:52

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Yes, for example, you can if (foo == FALSE) {opts_chunk$set(eval=FALSE, include=FALSE)}; then after this chunk, the global chunk options are changed so that the chunks are neither evaluated nor included in the output.

> Originally posted on 2013-02-21 00:38:51

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
No. The same labels in two Rnw files which are under the same directory will cause cache clash; only cache files for one chunk will survive.

> Originally posted on 2013-03-14 05:59:30

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
See <http://bit.ly/knitr-faq> #7

> Originally posted on 2013-03-30 15:53:50

**Reply by yihui** [reply]
Sorry for the confusion. The `width` option is a suggestion only. It is not strictly respected. If you want precise control over the code width, you need to use tidy=FALSE and knitr will not reformat the code.

> Originally posted on 2013-03-30 23:58:09

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sorry I cannot reproduce it. Please provide a reproducible example along with your sessionInfo().

> Originally posted on 2013-04-24 09:12:18

**Reply by yihui** [reply]
Either StackOverflow or Github issues will be OK; the latter is preferred, though. Thanks!

> Originally posted on 2013-05-25 07:05:31

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I'll fix it. Thanks!

> Originally posted on 2013-05-27 01:14:36

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
For the first question, it is a hard LaTeX problem, and there is no general solution.

For the second question, for now you only have cache=TRUE; it is currently on my TODO list: <https://github.com/yihui/knitr/issues/396>

> Originally posted on 2013-06-04 19:22:30

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
That will be too much work for me, and is exactly what fig.scap is for. Thanks!

> Originally posted on 2013-06-08 20:32:28

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Yes you can do that.

library(knitr)  
opts_chunk$set(message = FALSE)

The problem is you will load knitr for all R sessions, and that will slow R down by about 0.3 seconds.

> Originally posted on 2013-07-18 18:05:17

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
I think you can redefine the knitrout environment, which is empty by default and your use case is exactly why it was created.

> Originally posted on 2013-07-31 20:30:07

[a comment by an automated account is not shown]

**Comment by yihui** [comment]
If you want to hide everything, use include=FALSE; if you want to hide the messages, use message=FALSE. See <http://yihui.name/knitr/demo/output/>

> Originally posted on 2012-05-26 05:55:38

[a comment by an automated account is not shown]

**Comment by yihui** [comment]
Please read the documentation ?print.xtable carefully. There are no arguments named tabular.environments. I can only see tabular.environment, and I do not see it can take a vector of length 2 in its documentation.

The only relevant piece for knitr is you need to set the chunk option results='asis' for that chunk.

> Originally posted on 2012-06-19 14:39:27

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Thanks! Next time you can try to edit the Markdown file in the Github repo: <https://github.com/yihui/knitr/tree/gh-pages> (cf <http://yihui.name/en/2013/06/fix-typo-in-documentation/> )

> Originally posted on 2016-03-28 04:51:22

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
No short cut but you can use the dropdown menu "Insert" from the RStudio toolbar.

> Originally posted on 2017-12-14 23:07:08

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
You need to wait for the next version of knitr, or the current devel version, to be able to run Python code across chunks in the same Python session: <https://github.com/yihui/knitr/issues/1440>

> Originally posted on 2017-12-19 20:12:50

**Reply by yihui** [reply]
Please just update knitr to 1.18 and install the reticulate package in R.

> Originally posted on 2017-12-28 15:02:48

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
That sounds like an RStudio issue, which is beyond my expertise. Please consider asking the question on RStudio Community or file an issue to the Github repo rstudio/rstudio. Thanks!

> Originally posted on 2021-04-12 13:39:30

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
Sorry, I didn't get it. What do you mean?

> Originally posted on 2021-11-12 05:46:53

[a comment by an automated account is not shown]

**Reply by yihui** [reply]
This is not caused by knitr; knitr doesn't have control over the R console.

> Originally posted on 2021-11-10 14:25:09

[a comment by an automated account is not shown]

**Comment by courtiol** [comment]
FYI: in 
> size: ('normalsize'; character) Font size of the chunk output from .Rnw documents. See this page for possible sizes.

The URL link is broken. Perhaps this is temporary(?), in the meantime, here is another page covering the topic: https://latex-tutorial.com/changing-font-size/

**Reply by yihui** [reply]
Thanks for the reminder! I can access the link right now. Perhaps it was a temporary problem.

**Comment by drag05** [comment]
Hello yihui

I am rendering a pdf document from within Shiny app and at one instance I need to include a png plot previously saved. The Rmd file is placed in the root directory of the shiny app along with global.R, server.R and ui.R. Function `includegraphics()` works fine if I use it inline but not so inside code chunk:

Inline, but with no caption (caption throws a "float" error in this case):

```r
knitr::include_graphics('./src/coverage001.png', dpi = 72, auto_pdf = TRUE)
```

Inside code chunk preferably with caption:

````md
```{r covPlot, fig.align='center',  fig.ncol = 1, fig.sep = '\\par', fig.width = 9, fig.show = 'asis', fig.keep = 'all', fig.cap = sprintf('Protein %s: Coverage', names(chain))}
# coverage plot
knitr::include_graphics('./src/coverage001.png', dpi = 72, auto_pdf = TRUE)
```
````

 When I run the chunk in RStudio, the figure appears under the chunk but not so inside the rendered document. No latex error either in the log file. I am using `graphicx` package without any success and all R packages are at their latest version. I have also `tlmgr_reinstalled tinytex and updated, to no avail. Please advise, thank you!

**Reply by yihui** [reply]
> the figure appears under the chunk but not so inside the rendered document

I'm not sure what that means. Do you know that figure environments in LaTeX float by default? That is, usually they don't appear exactly where you place them in the source document. Reference: https://bookdown.org/yihui/rmarkdown-cookbook/figure-placement.html

**Comment by drag05** [comment]
@yihui  Thank you for the prompt reply!
 Is not about *where* the figure appears in the pdf document. My question was referring to *why* the figure only appears in the pdf document when `include_graphics` is used inline and why it does not appear  *at all*  in the pdf document when `include graphics` is used inside the code chunk. 

In other words, what should I add to the chunk options to have the figure appear when I use `include_graphics` inside the code chunk (please see chunk options in example 2).

This is what I mean by inline (example 1):

```
`r knitr::include_graphics('./src/coverage001.png', dpi = 72, auto_pdf = TRUE)`
```

and this is what I mean by code chunk (example 2):

````
```{r  fig.align='center',  fig.ncol = 1, fig.sep = '\\par', fig.width = 9, fig.show = 'asis', fig.keep = 'all', fig.cap = "A00B10"}
knitr::include_graphics('./src/coverage001.png', dpi = 72, auto_pdf = TRUE)
```
````

This is part of my document header, including latex packages (the indents are stripped by the browser - they are correct in the Rmd):

```
output: 
  pdf_document: 
    latex_engine: xelatex
    df_print: paged
    fig_caption: yes
    toc: yes
    extra_dependencies: float
    highlight: tango
lof: yes
lot: yes
fontawesome: yes
fig.align: center
graphics: yes
header-includes:
 \usepackage{fontspec} \usepackage{graphicx} \usepackage{pdflscape} \usepackage{rotating} \usepackage{titling} \setlength{\droptitle}{-1cm} \usepackage{float} \let\origfigure\figure \let\endorigfigure\endfigure \renewenvironment{figure}[1][2] {\expandafter\origfigure\expandafter[H]} {\endorigfigure} \usepackage{mdframed} \usepackage{subfig} \usepackage{fancyhdr} \setlength{\headheight}{42.45442pt} \addtolength{\topmargin}{-1.3pt} \usepackage{graphicx} \usepackage{booktabs,xcolor} \usepackage{lastpage} \usepackage{longtable} \pagestyle{fancy} \fancyhf{} \fancyfoot[R]{\thepage} \setlength\footskip{14.5pt} \fancypagestyle{plain}{\pagestyle{fancy}} \lhead{\today}  \rhead{\includegraphics[width=2.5cm]{./www/logo.png}}
mainfont: Palatino Linotype
geometry: margin = 0.75in
fontsize: 11pt
```

Regarding latex packages: where can I find some *succinct* information on what latex package(s) to call with `\use{}` for various tasks?

Thank you!

**Reply by yihui** [reply]
> My question was referring to _why_ the figure only appears in the pdf document when `include_graphics` is used inline and why it does not appear _at all_ in the pdf document when `include graphics` is used inside the code chunk.

I see. That's odd. Sorry I have no idea... A minimal reproducible example will be helpful.

**Comment by guanghu52** [comment]
I had a render pdf problem when my images had multiple dots in the filename, knitr.include.graphics.ext not run?
